File: backend/app/local/claude_oauth.py
"""OAuth da assinatura Claude Code. No macOS o token vive no Keychain."""

from __future__ import annotations

import json
import os
import subprocess
import time
from pathlib import Path
from typing import Any

KEYCHAIN_SERVICES = ("Claude Code-credentials",)
KEYCHAIN_CACHE_TTL_S = 30.0

_last_keychain_err: str | None = None
_keychain_cache: tuple[float, tuple[str | None, int | None, str | None]] | None = None


def parse_oauth_blob(data: Any) -> tuple[str | None, int | None]:
    if isinstance(data, str):
        try:
            data = json.loads(data)
        except json.JSONDecodeError:
            token = data.strip()
            return (token, None) if token else (None, None)
    if not isinstance(data, dict):
        return None, None
    oauth = data.get("claudeAiOauth") or data.get("oauth") or data
    if not isinstance(oauth, dict):
        return None, None
    token = (oauth.get("accessToken") or data.get("accessToken") or "").strip()
    exp = oauth.get("expiresAt") or data.get("expiresAt")
    exp_i = int(exp) if isinstance(exp, (int, float)) else None
    if not token:
        return None, None
    return token, exp_i


def from_credentials_file(path: Path) -> tuple[str | None, int | None, str | None]:
    if not path.is_file():
        return None, None, None
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as exc:
        return None, None, f"credentials.json: {exc}"
    token, exp = parse_oauth_blob(data)
    if not token:
        return None, None, "credentials.json sem accessToken"
    return token, exp, None


def _keychain_services() -> list[str]:
    names = list(KEYCHAIN_SERVICES)
    extra = os.environ.get("CLAUDE_KEYCHAIN_SERVICE", "").strip()
    if extra and extra not in names:
        names.insert(0, extra)
    return names


def _keychain_error(svc: str, proc: subprocess.CompletedProcess[str]) -> str:
    raw = (proc.stderr or proc.stdout or "").strip()
    low = raw.lower()
    if "could not be found" in low or proc.returncode == 44:
        return (
            f"Keychain sem o item «{svc}». Abra o Claude Code (`claude`) neste Mac, "
            "faça login, e se o sistema pedir, permita o acesso do coletor."
        )
    if "user interaction is not allowed" in low:
        return (
            "Keychain recusou (sem janela para confirmar). Rode o coletor no Terminal "
            "deste Mac — não no Docker — e permita o acesso quando o macOS pedir."
        )
    if "errsecauthfailed" in low or "authorization" in low:
        return "Keychain recusou a senha. Rode `claude` e tente o coletor de novo no Terminal."
    return (raw or "keychain recusou")[:240]


def last_keychain_error() -> str | None:
    return _last_keychain_err


def from_macos_keychain() -> tuple[str | None, int | None, str | None]:
    global _last_keychain_err, _keychain_cache
    if _keychain_cache is not None:
        cached_at, result = _keychain_cache
        if time.monotonic() - cached_at < KEYCHAIN_CACHE_TTL_S:
            _last_keychain_err = result[2]
            return result
    _last_keychain_err = None
    if os.uname().sysname != "Darwin":
        return None, None, None
    best: tuple[str, int | None] | None = None
    last_err = None
    for svc in _keychain_services():
        try:
            proc = subprocess.run(
                ["security", "find-generic-password", "-s", svc, "-w"],
                check=False,
                capture_output=True,
                text=True,
                timeout=20,
            )
        except (OSError, subprocess.TimeoutExpired) as exc:
            last_err = str(exc)
            continue
        if proc.returncode != 0:
            last_err = _keychain_error(svc, proc)
            continue
        token, exp = parse_oauth_blob(proc.stdout.strip())
        if not token:
            last_err = f"item {svc} sem accessToken"
            continue
        if best is None or (exp or 0) >= (best[1] or 0):
            best = (token, exp)
    if best:
        _last_keychain_err = None
        result = (best[0], best[1], None)
    else:
        _last_keychain_err = last_err
        result = (None, None, last_err)
    _keychain_cache = (time.monotonic(), result)
    return result


def credentials_path(cfg: dict | None = None) -> Path:
    env = os.environ.get("CLAUDE_CREDENTIALS_PATH", "").strip()
    if env:
        return Path(env).expanduser()
    stored = ""
    if cfg:
        stored = str((cfg.get("paths") or {}).get("claude_credentials") or "").strip()
    if stored:
        return Path(stored).expanduser()
    return Path.home() / ".claude" / ".credentials.json"


def claude_token_candidates(cfg: dict | None = None) -> list[tuple[str, str, int | None]]:
    """Keychain e credentials.json. Token colado no painel não entra aqui."""
    found: list[tuple[str, str, int | None]] = []
    seen: set[str] = set()

    def add(source: str, token: str | None, exp: int | None) -> None:
        if not token or token in seen:
            return
        seen.add(token)
        found.append((source, token, exp))

    kc_token, kc_exp, _kc_err = from_macos_keychain()
    add("keychain", kc_token, kc_exp)
    file_tok, file_exp, _err = from_credentials_file(credentials_path(cfg))
    add("credentials", file_tok, file_exp)
    return found

File: backend/src/local/claudeOauth.ts
/**
 * OAuth da assinatura Claude Code. No macOS o token vive no Keychain.
 * Port of backend-python-legacy/app/local/claude_oauth.py
 */
import { spawnSync } from "node:child_process";
import { existsSync, readFileSync } from "node:fs";
import { homedir, platform } from "node:os";
import { join } from "node:path";

const KEYCHAIN_SERVICES: readonly string[] = ["Claude Code-credentials"] as const;
const KEYCHAIN_CACHE_TTL_S = 30.0;

let _lastKeychainErr: string | null = null;
let _keychainCache: { at: number; result: [string | null, number | null, string | null] } | null = null;

export function parseOauthBlob(data: unknown): [string | null, number | null] {
  if (typeof data === "string") {
    try {
      data = JSON.parse(data);
    } catch {
      const token = (data as string).trim();
      return token ? [token, null] : [null, null];
    }
  }
  if (data === null || typeof data !== "object" || Array.isArray(data)) {
    return [null, null];
  }
  const dict = data as Record<string, unknown>;
  const oauthRaw = (dict.claudeAiOauth ?? dict.oauth ?? dict) as unknown;
  if (oauthRaw === null || typeof oauthRaw !== "object" || Array.isArray(oauthRaw)) {
    return [null, null];
  }
  const oauth = oauthRaw as Record<string, unknown>;
  const token = String(oauth.accessToken ?? dict.accessToken ?? "").trim();
  const expRaw = oauth.expiresAt ?? dict.expiresAt;
  // Python: int(exp) if isinstance(exp, (int, float)) else None
  const expI: number | null = typeof expRaw === "number" && Number.isFinite(expRaw) ? Math.trunc(expRaw) : null;

  if (!token) return [null, null];
  return [token, expI];
}

export function fromCredentialsFile(path: string): [string | null, number | null, string | null] {
  if (!existsSync(path)) return [null, null, null];
  let data: unknown;
  try {
    data = JSON.parse(readFileSync(path, "utf-8"));
  } catch (e) {
    return [null, null, `credentials.json: ${e}`];
  }
  const [token, exp] = parseOauthBlob(data);
  if (!token) return [null, null, "credentials.json sem accessToken"];
  return [token, exp, null];
}

function keychainServices(): string[] {
  const names = [...KEYCHAIN_SERVICES];
  const extra = (process.env.CLAUDE_KEYCHAIN_SERVICE ?? "").trim();
  if (extra && !names.includes(extra)) names.unshift(extra);
  return names;
}

function keychainError(svc: string, proc: ReturnType<typeof spawnSync>): string {
  const raw = String(proc.stderr ?? proc.stdout ?? "").trim();
  const low = raw.toLowerCase();
  if (low.includes("could not be found") || proc.status === 44) {
    return (
      `Keychain sem o item «${svc}». Abra o Claude Code (\`claude\`) neste Mac, ` +
      "faça login, e se o sistema pedir, permita o acesso do coletor."
    );
  }
  if (low.includes("user interaction is not allowed")) {
    return (
      "Keychain recusou (sem janela para confirmar). Rode o coletor no Terminal " +
      "deste Mac — não no Docker — e permita o acesso quando o macOS pedir."
    );
  }
  if (low.includes("errsecauthfailed") || low.includes("authorization")) {
    return "Keychain recusou a senha. Rode `claude` e tente o coletor de novo no Terminal.";
  }
  return (raw || "keychain recusou").slice(0, 240);
}

export function lastKeychainError(): string | null {
  return _lastKeychainErr;
}

export function fromMacosKeychain(): [string | null, number | null, string | null] {
  if (_keychainCache !== null) {
    const { at, result } = _keychainCache;
    if (Date.now() / 1000 - at < KEYCHAIN_CACHE_TTL_S) {
      _lastKeychainErr = result[2];
      return result;
    }
  }
  _lastKeychainErr = null;
  if (platform() !== "darwin") return [null, null, null];

  let best: [string, number | null] | null = null;
  let lastErr: string | null = null;

  for (const svc of keychainServices()) {
    let proc: ReturnType<typeof spawnSync>;
    try {
      proc = spawnSync("security", ["find-generic-password", "-s", svc, "-w"], {
        encoding: "utf-8",
        timeout: 20_000,
      });
    } catch (e) {
      lastErr = String(e);
      continue;
    }
    if (proc.error) {
      lastErr = String(proc.error);
      continue;
    }
    if (proc.status !== 0) {
      lastErr = keychainError(svc, proc);
      continue;
    }
    const out = String(proc.stdout ?? "").trim();
    const [token, exp] = parseOauthBlob(out);
    if (!token) {
      lastErr = `item ${svc} sem accessToken`;
      continue;
    }
    if (best === null || (exp ?? 0) >= (best[1] ?? 0)) {
      best = [token, exp];
    }
  }

  let result: [string | null, number | null, string | null];
  if (best) {
    _lastKeychainErr = null;
    result = [best[0], best[1], null];
  } else {
    _lastKeychainErr = lastErr;
    result = [null, null, lastErr];
  }
  _keychainCache = { at: Date.now() / 1000, result };
  return result;
}

function expandUser(p: string): string {
  if (p === "~") return homedir();
  if (p.startsWith("~/")) return join(homedir(), p.slice(2));
  return p;
}

export function credentialsPath(cfg?: Record<string, unknown> | null): string {
  const env = (process.env.CLAUDE_CREDENTIALS_PATH ?? "").trim();
  if (env) return expandUser(env);
  let stored = "";
  if (cfg) {
    const paths = cfg.paths as Record<string, unknown> | undefined;
    stored = String(paths?.claude_credentials ?? "").trim();
  }
  if (stored) return expandUser(stored);
  return join(homedir(), ".claude", ".credentials.json");
}

export function missingLoginHint(cfg?: Record<string, unknown> | null): string {
  if (platform() === "darwin") {
    return lastKeychainError() ?? "Nenhum login encontrado — rode `claude` neste Mac";
  }
  return (
    "Nenhum login encontrado — rode `claude` neste computador " + `(o token fica em ${credentialsPath(cfg)})`
  );
}

export function claudeTokenCandidates(
  cfg?: Record<string, unknown> | null,
): Array<[string, string, number | null]> {
  const found: Array<[string, string, number | null]> = [];
  const seen = new Set<string>();

  function add(source: string, token: string | null, exp: number | null): void {
    if (!token || seen.has(token)) return;
    seen.add(token);
    found.push([source, token, exp]);
  }

  const [kcToken, kcExp] = fromMacosKeychain();
  add("keychain", kcToken, kcExp);
  const [fileTok, fileExp] = fromCredentialsFile(credentialsPath(cfg));
  add("credentials", fileTok, fileExp);
  return found;
}

// For testing: reset cache
export function _resetClaudeOauthCache(): void {
  _keychainCache = null;
  _lastKeychainErr = null;
}
